Fix diamond inlining of rand

bertmaher opened the pull request.

bertmaher committed:
Fix diamond inlining of rand

zheng-xq commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
+ */
+class RandomInliner : public IRMutator {
+ public:
+  RandomInliner(const std::unordered_set<Function*>& funcs) : funcs_(funcs) {
```
explicit?

zheng-xq commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
+/**
+ * Actual algorithm for "inlining" of rand().  Bind variables to rand() calls.
+ */
+class RandomInliner : public IRMutator {
```
I'm wondering whether RandomInliner should inherit from FunctionInliner, and extend the additional functionalities.

bertmaher replied:
Yeah I think it should.  I started off by copying the impl on the assumption it would be more different, but it's really not.

zheng-xq commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
+  // TODO: Test this with multiple different random functions.
+  //
+  // TODO: Can we have accesses to the same random tensor with different
+  // indices?  We should probably generate different random numbers for those.
```
In that case, that particular inlining should be disabled. For example: 

B(i, j) = A(i, j) + A(i+1, j), where A(i, j) = random()

For the correct implementation, B(i, j) and B(i-1, j) have a strong spatial correlation. But if implemented as "rand() + rand()", that spatial correctional goes away.

bertmaher replied:
This is a great point, I was assuming nonoverlapping dependences.  I think we probably just want to track the call indices and bail out if there are any mismatches.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
+
+  static bool is_inner_loop(const For* loop) {
+    Block* body = dynamic_cast<Block*>(loop->body());
+    if (!body) {
```
This should be turned to an assert/removed. Bodies of For loops are blocks by construction.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
+    if (!body) {
+      return false;
+    }
+    if (body->nstmts() != 1) {
```
This seems to be overly conservative. It should be allowed to have several statements in the body if neither of them is a For loop:
```
for // <-- loop
  block // <-- body
    store1 // <-- *body->stmts().begin()
    store2 // <-- *(body->stmts().begin()++)
```

ZolotukhinM commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
+      return new For(var, start, stop, new_body, loop_options);
+    }
+
+    Stmt* new_body = body->accept_mutator(this);
```
You need to clone stmt before mutating.

bertmaher replied:
Why?  I don't think I actually need a copy of the old body?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
+    Stmt* new_body = body->accept_mutator(this);
+    for (auto const& p : random_vars_) {
+      Var* v = p.second;
+      new_body = new LetStmt(v, new Intrinsics(kRand, v->dtype()), new_body);
```
It would be helpful to add a comment describing how we're inserting the let statement and why this helps (maybe as a class comment).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
+    Function* prev_func = current_func_;
+    current_func_ = func;
+    if (!random_vars_.count(func)) {
+      random_vars_[func] = new Var("v", v->dtype());
```
Nit: maybe use the function's name for the variable name instead of "v".

ZolotukhinM commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
+    if (v->op_type() != kRand) {
+      return v;
+    }
+    return random_vars_[current_func_];
```
Nit: `s/random_vars_[current_func_]/random_vars_.at(current_func_)/`

ZolotukhinM reviewed (commented):
I also feel that it's a bit hacky, but if it works I prefer to have it rather than wait for an ideal implementation. I didn't dig deep into the transformation logic itself.

bertmaher committed:
Fixes

bertmaher commented:
> I also feel that it's a bit hacky

Yeah I mean I'm not really sure if there's a better alternative.  Do you have any suggestion how to make this sort of approach not a hack? (The permutation polynomial idea from Halide is one possibility but I don't totally understand it.)

bertmaher committed:
improvements from MikeZ

bertmaher committed:
clang-format

bertmaher committed:
Fix some corner cases

bertmaher committed:
clang-format

ZolotukhinM commented on test/test_jit_fuser.py:
```diff
@@ -860,7 +863,6 @@ def fn_test_rand2(x, y):
         out = script_f(x, y)
         self.assertEqual(out[0, :] + torch.zeros(4, 4, device='cuda'), out)
 
-    @unittest.skip("TE fuser broken")
```
Please add`@unittest.skipIf(not RUN_CUDA, "fuser requires CUDA")` since we're using cuda_codegen triggers here

ZolotukhinM commented on test/test_tensorexpr.py:
```diff
@@ -1225,6 +1225,15 @@ def run_where(x, y):
         y = run_where(a, b)
         np.testing.assert_allclose(x.numpy(), y.numpy())
 
+    def test_unused(self):
```
Ditto.

zheng-xq replied:
We should also add a function test: 

A(i, j) = rand_like()
B = (2 * A - 1) * (2 * A - 2)
assert(B(i, j) >= -0.25)

If both A are inlined as rand, this is going to fail. But with your fix, this is going to pass.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,132 @@ class FunctionInliner : public IRMutator {
+    if (body->nstmts() != 1) {
+      return true;
+    }
+    return dynamic_cast<For*>(*body->stmts().begin()) == nullptr;
```
The check seems to be incorrect: it basically says "all loops with more than one inner statement, the first one of which is not For, are innermost loops".  A counterexample:
```
for i
  a[i] = b[i]
  for j
    c[i,j] = 0
```

The correct condition, I think, is `!any_of(body->stmts(), is<For>)` (pls forgive my pseudo code).

bertmaher replied:
Yes, you have a good point here.  I'm being flummoxed by this because I'm essentially just trying to get past the loops we've created as part of lowering.  But between this and XQ's comment about if-then-else, I think I actually need to do something a bit trickier, which is to "bubble up" the rand to its nearest containing control structure (whether that's a loop or an if-then-else.

I seriously don't understand why this is so hard!  Maybe I will go study Halide's implementation while I am off on parental leave next week :-p

zheng-xq commented on test/test_jit_fuser.py:
```diff
@@ -796,7 +796,10 @@ def create(self, x):
         x = torch.zeros([3, 4, 5], dtype=torch.float, device='cuda')
         m = M()
         out1 = m.create(x)
+        start_execs = torch._C._jit_get_trigger_value("cuda_codegen_executed")
```
Maybe add a TODO to refactor CudaCodeGenExecuted and use it here?

bertmaher replied:
Sure, let me go ahead and do that in a follow up.

zheng-xq commented on torch/csrc/jit/tensorexpr/schedule.h:
```diff
@@ -22,10 +22,12 @@ class TORCH_API LoopNest {
 
   void Vectorize(Stmt*);
   void ComputeInline(Stmt* s);
+  void ComputeInlineWithRandom(Stmt* s);
```
Minor: I am a bit split whether this should be exposed as part of the Schedule, or as part of implementation details within ComputeInline.

bertmaher replied:
I had this thought too.  Would be open to changing it down the line.

zheng-xq commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,132 @@ class FunctionInliner : public IRMutator {
+      : FunctionInliner(funcs) {}
+
+  // Find the innermost for loop and bind any random variables in its body.
+  Stmt* mutate(const For* v) override {
```
I am wondering whether we should respect other common C++ scopes: conditionals, etc. 

It seems wrong if we move an evaluation outside a if-then-else.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/schedule.cpp:
```diff
@@ -399,6 +399,132 @@ class FunctionInliner : public IRMutator {
+    }
+    auto it = random_vars_.find(current_func_);
+    if (it == random_vars_.end()) {
+      return v;
```
Should this ever happen? Can we replace it with an assert checking that all `rand` calls are replaced with a var+let_stmt?

bertmaher replied:
Yeah, I think this can actually happen if you choose not to inline some functions containing rand() (not that it happens with our current fuser, but still)

ZolotukhinM reviewed (approved):
Thanks for addressing my previous comments! Overall this looks good to me modulo several remarks (see inline).

bertmaher committed:
Easy fixes plus corner cases

bertmaher committed:
Bind random vars in innermost control structure (if-then-else or for-loop)

bertmaher committed:
clang-format

bertmaher committed:
Placate clang-tidy regarding const_cast

bertmaher commented:
Thanks for the detailed reviews, guys!  I think I've addressed all the concerns (esp the one about finding innermost loops/control structures), so I'm going to go ahead and land this.
